File: typescript/tests/voice_listener.test.ts
// voice_listener.test.ts
//
// Verifies VoiceCompanionListener (VoiceCompanionListener.cs): a transcription
// raises UtteranceDetected, is forwarded to the session, and the reply raises
// ResponseReady; start/stop drive the pipeline; disposeAsync unsubscribes and
// disposes both the pipeline and the session; a session error does not crash.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  VoiceCompanionListener,
  type IVoicePipeline,
  type TranscribedHandler,
  type TranscribedEventArgs,
} from '../src/companion/index';
import type { ICompanionSession } from '../src/companion/index';

/** A controllable fake pipeline that lets the test push a transcription. */
class FakePipeline implements IVoicePipeline {
  private handlers = new Set<TranscribedHandler>();
  started = false;
  stopped = false;
  disposed = false;

  onTranscribed(h: TranscribedHandler): void {
    this.handlers.add(h);
  }
  offTranscribed(h: TranscribedHandler): void {
    this.handlers.delete(h);
  }
  async startAsync(): Promise<void> {
    this.started = true;
  }
  async stopAsync(): Promise<void> {
    this.stopped = true;
  }
  async disposeAsync(): Promise<void> {
    this.disposed = true;
  }
  emit(args: TranscribedEventArgs): void {
    for (const h of [...this.handlers]) h(args);
  }
  get handlerCount(): number {
    return this.handlers.size;
  }
}

function fakeSession(reply: string | (() => Promise<string>)): ICompanionSession & {
  disposeAsync: () => Promise<void>;
  disposed: boolean;
} {
  const state = { disposed: false };
  return {
    sessionId: 's',
    identityId: 'i',
    interface: 0 as never,
    onProactiveMessageReady: null,
    history: [],
    async sendAsync(): Promise<string> {
      return typeof reply === 'function' ? reply() : reply;
    },
    async *streamAsync() {
      /* unused */
    },
    async agentAsync() {
      return '';
    },
    getContext() {
      return {} as never;
    },
    async refreshContextAsync() {
      /* unused */
    },
    async signalFeedbackAsync() {
      /* unused */
    },
    async disposeAsync() {
      state.disposed = true;
    },
    get disposed() {
      return state.disposed;
    },
  } as unknown as ICompanionSession & { disposeAsync: () => Promise<void>; disposed: boolean };
}

function transcription(text: string, confidence = 0.9): TranscribedEventArgs {
  return { result: { text, confidence }, completedAt: new Date() };
}

describe('VoiceCompanionListener', () => {
  it('raises UtteranceDetected then ResponseReady with the reply', async () => {
    const pipeline = new FakePipeline();
    const session = fakeSession('the reply');
    const listener = new VoiceCompanionListener(pipeline, session);

    const utterances: string[] = [];
    const responses: { text: string; original: string }[] = [];
    listener.onUtteranceDetected((a) => utterances.push(a.text));
    listener.onResponseReady((a) => responses.push({ text: a.text, original: a.originalUtterance }));

    pipeline.emit(transcription('what time is it'));
    // ResponseReady fires from an async continuation; give it a tick.
    await new Promise((r) => setTimeout(r, 5));

    assert.deepEqual(utterances, ['what time is it']);
    assert.deepEqual(responses, [{ text: 'the reply', original: 'what time is it' }]);
    await listener.disposeAsync();
  });

  it('start/stop drive the underlying pipeline', async () => {
    const pipeline = new FakePipeline();
    const listener = new VoiceCompanionListener(pipeline, fakeSession('x'));
    await listener.startAsync();
    assert.equal(pipeline.started, true);
    await listener.stopAsync();
    assert.equal(pipeline.stopped, true);
    await listener.disposeAsync();
  });

  it('disposeAsync unsubscribes and disposes pipeline + session', async () => {
    const pipeline = new FakePipeline();
    const session = fakeSession('x');
    const listener = new VoiceCompanionListener(pipeline, session);
    assert.equal(pipeline.handlerCount, 1);
    await listener.disposeAsync();
    assert.equal(pipeline.handlerCount, 0);
    assert.equal(pipeline.disposed, true);
    assert.equal((session as unknown as { disposed: boolean }).disposed, true);
    // start after dispose throws.
    await assert.rejects(() => listener.startAsync(), /disposed/);
  });

  it('a session error does not crash the listener (no ResponseReady)', async () => {
    const pipeline = new FakePipeline();
    const session = fakeSession(async () => {
      throw new Error('model down');
    });
    const listener = new VoiceCompanionListener(pipeline, session);
    const responses: string[] = [];
    listener.onResponseReady((a) => responses.push(a.text));
    pipeline.emit(transcription('hello'));
    await new Promise((r) => setTimeout(r, 5));
    assert.equal(responses.length, 0); // failure swallowed
    await listener.disposeAsync();
  });

  it('rejects null pipeline / session in the constructor', () => {
    // @ts-expect-error deliberate null pipeline
    assert.throws(() => new VoiceCompanionListener(null, fakeSession('x')), /pipeline required/);
  });
});

File: rust/tests/voice_listener_test.rs
//! voice_listener_test.rs
//!
//! Verifies VoiceCompanionListener: a transcription raises UtteranceDetected,
//! forwards to the session-send closure, and raises ResponseReady with the reply.
//! Session failures never raise ResponseReady. Mirrors the C#
//! VoiceCompanionListener behaviour.

use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};

use chrono::Utc;
use circle_ai::companion::voice_listener::*;

/// A pipeline that just tracks start/stop.
struct FakePipeline {
    started: Arc<AtomicBool>,
    stopped: Arc<AtomicBool>,
}
impl IVoicePipeline for FakePipeline {
    fn start(&self) {
        self.started.store(true, Ordering::SeqCst);
    }
    fn stop(&self) {
        self.stopped.store(true, Ordering::SeqCst);
    }
}

#[test]
fn transcription_forwards_and_raises_response() {
    let started = Arc::new(AtomicBool::new(false));
    let stopped = Arc::new(AtomicBool::new(false));
    let pipeline = Arc::new(FakePipeline {
        started: started.clone(),
        stopped: stopped.clone(),
    });
    let send: SessionSendFn = Arc::new(|text: &str| Ok(format!("reply to: {text}")));
    let listener = VoiceCompanionListener::new(pipeline, send);

    let utterances = Arc::new(Mutex::new(Vec::<String>::new()));
    let responses = Arc::new(Mutex::new(Vec::<(String, String)>::new()));
    let u2 = utterances.clone();
    let r2 = responses.clone();
    listener.on_utterance_detected(Box::new(move |e| u2.lock().unwrap().push(e.text.clone())));
    listener.on_response_ready(Box::new(move |e| {
        r2.lock()
            .unwrap()
            .push((e.original_utterance.clone(), e.text.clone()))
    }));

    listener.start();
    assert!(started.load(Ordering::SeqCst));

    listener.on_transcribed(
        TranscriptionResult {
            text: "hey b, what's the time".into(),
            confidence: 0.92,
        },
        Utc::now(),
    );

    assert_eq!(utterances.lock().unwrap().len(), 1);
    let resp = responses.lock().unwrap();
    assert_eq!(resp.len(), 1);
    assert_eq!(resp[0].0, "hey b, what's the time");
    assert_eq!(resp[0].1, "reply to: hey b, what's the time");
}

#[test]
fn session_failure_raises_utterance_but_not_response() {
    let pipeline = Arc::new(FakePipeline {
        started: Arc::new(AtomicBool::new(false)),
        stopped: Arc::new(AtomicBool::new(false)),
    });
    let send: SessionSendFn = Arc::new(|_| Err("model crashed".to_string()));
    let listener = VoiceCompanionListener::new(pipeline, send);

    let utter_count = Arc::new(AtomicUsize::new(0));
    let resp_count = Arc::new(AtomicUsize::new(0));
    let u2 = utter_count.clone();
    let r2 = resp_count.clone();
    listener.on_utterance_detected(Box::new(move |_| {
        u2.fetch_add(1, Ordering::SeqCst);
    }));
    listener.on_response_ready(Box::new(move |_| {
        r2.fetch_add(1, Ordering::SeqCst);
    }));

    listener.on_transcribed(
        TranscriptionResult {
            text: "hi".into(),
            confidence: 1.0,
        },
        Utc::now(),
    );

    // Utterance detected fired; response ready did not (failure swallowed).
    assert_eq!(utter_count.load(Ordering::SeqCst), 1);
    assert_eq!(resp_count.load(Ordering::SeqCst), 0);
}

#[test]
fn disposed_listener_ignores_transcriptions_and_stops_pipeline() {
    let stopped = Arc::new(AtomicBool::new(false));
    let pipeline = Arc::new(FakePipeline {
        started: Arc::new(AtomicBool::new(false)),
        stopped: stopped.clone(),
    });
    let send: SessionSendFn = Arc::new(|t: &str| Ok(t.to_string()));
    let listener = VoiceCompanionListener::new(pipeline, send);

    let resp_count = Arc::new(AtomicUsize::new(0));
    let r2 = resp_count.clone();
    listener.on_response_ready(Box::new(move |_| {
        r2.fetch_add(1, Ordering::SeqCst);
    }));

    listener.dispose();
    assert!(stopped.load(Ordering::SeqCst));

    listener.on_transcribed(
        TranscriptionResult {
            text: "after dispose".into(),
            confidence: 1.0,
        },
        Utc::now(),
    );
    assert_eq!(resp_count.load(Ordering::SeqCst), 0);
}

#[test]
fn start_stop_drive_the_pipeline() {
    let started = Arc::new(AtomicBool::new(false));
    let stopped = Arc::new(AtomicBool::new(false));
    let pipeline = Arc::new(FakePipeline {
        started: started.clone(),
        stopped: stopped.clone(),
    });
    let send: SessionSendFn = Arc::new(|t: &str| Ok(t.to_string()));
    let listener = VoiceCompanionListener::new(pipeline, send);
    listener.start();
    listener.stop();
    assert!(started.load(Ordering::SeqCst));
    assert!(stopped.load(Ordering::SeqCst));
}
